plumbing/reference.go

```diff
@@ -205,9 +222,14 @@ func (r ReferenceName) Validate() error {
 			return ErrInvalidReferenceName
 		}
 
-		if (isBranch || isTag) && strings.HasPrefix(part, "-") { // branches & tags can't start with -
-			return ErrInvalidReferenceName
-		}
+	}
+
+	if (isBranch) && strings.HasPrefix(r.BranchName(), "-") { // branches & tags can't start with -
+		return ErrInvalidReferenceName
+	}
+
+	if (isTag) && strings.HasPrefix(r.TagName(), "-") { // branches & tags can't start with -
+		return ErrInvalidReferenceName
```

Comment:
This can be simplified by keeping the existing if and just adding ` && i == 2` to it:
```suggestion
		if (isBranch || isTag) && strings.HasPrefix(part, "-") && i == 2 { // branches & tags can't start with -
			return ErrInvalidReferenceName
		}
```

This will need the for on `L208` to change to `for i, part := range parts {`.